File: plugins/lorepia-platform/android/src/main/java/dev/lorepia/tauri/platform/PlatformPolicy.kt
package dev.lorepia.tauri.platform

import java.security.MessageDigest

internal object PlatformPolicy {
    const val MAXIMUM_REFERENCE_BYTES = 256
    const val MAXIMUM_CREDENTIAL_READ_BYTES = 32 * 1024
    const val MAXIMUM_CREDENTIAL_WRITE_BYTES = 16 * 1024
    const val MAXIMUM_IMPORT_BYTES = 50L * 1024L * 1024L
    const val COPY_BUFFER_BYTES = 64 * 1024
    const val MAXIMUM_DISPLAY_NAME_CHARACTERS = 255
    const val OWNED_STAGING_PREFIX = "lorepia-tauri-"
    const val ABANDONED_STAGING_AGE_MILLIS = 24L * 60L * 60L * 1_000L

    fun validateReference(reference: String) {
        require(reference.isNotBlank()) { "invalid reference" }
        val encoded = reference.toByteArray(Charsets.UTF_8)
        try {
            require(encoded.size <= MAXIMUM_REFERENCE_BYTES) { "invalid reference" }
        } finally {
            encoded.fill(0)
        }
    }

    fun validateCredentialForWrite(value: String): ByteArray {
        require(value.isNotBlank()) { "invalid credential" }
        return value.toByteArray(Charsets.UTF_8).also {
            require(it.size <= MAXIMUM_CREDENTIAL_WRITE_BYTES) {
                it.fill(0)
                "invalid credential"
            }
        }
    }

    fun validateCredentialForRead(value: ByteArray) {
        require(value.isNotEmpty() && value.size <= MAXIMUM_CREDENTIAL_READ_BYTES) {
            "invalid credential"
        }
    }

    fun credentialFileName(reference: String): String {
        val referenceBytes = reference.toByteArray(Charsets.UTF_8)
        val digest = try {
            MessageDigest.getInstance("SHA-256").digest(referenceBytes)
        } finally {
            referenceBytes.fill(0)
        }
        return try {
            buildString(digest.size * 2 + CREDENTIAL_SUFFIX.length) {
                for (byte in digest) {
                    append(HEX[(byte.toInt() ushr 4) and 0x0f])
                    append(HEX[byte.toInt() and 0x0f])
                }
                append(CREDENTIAL_SUFFIX)
            }
        } finally {
            digest.fill(0)
        }
    }

    fun sanitizeDisplayName(value: String?): String {
        val sanitized = value
            ?.take(MAXIMUM_DISPLAY_NAME_CHARACTERS)
            ?.map { character -> if (character.isISOControl()) '\uFFFD' else character }
            ?.joinToString(separator = "")
            ?.takeIf(String::isNotBlank)
        return sanitized ?: "selected-file"
    }

    fun stagingSuffix(displayName: String): String =
        when (displayName.substringAfterLast('.', "").lowercase()) {
            "charx" -> ".charx"
            "json" -> ".json"
            "zip" -> ".zip"
            else -> ".pending"
        }

    fun shouldRemoveAbandonedStagingFile(
        name: String,
        isRegularFile: Boolean,
        lastModifiedMillis: Long,
        nowMillis: Long,
    ): Boolean =
        name.startsWith(OWNED_STAGING_PREFIX) &&
            isRegularFile &&
            lastModifiedMillis > 0L &&
            nowMillis >= lastModifiedMillis &&
            nowMillis - lastModifiedMillis >= ABANDONED_STAGING_AGE_MILLIS

    private const val CREDENTIAL_SUFFIX = ".credential"
    private const val HEX = "0123456789abcdef"
}

File: plugins/lorepia-platform/ios/Sources/PlatformPolicy.swift
import Foundation

enum PlatformPolicyError: Error {
    case invalidCredential
    case invalidReference
}

enum PlatformPolicy {
    static let maximumReferenceBytes = 256
    static let maximumCredentialBytes = 16 * 1_024
    static let maximumImportBytes: UInt64 = 128 * 1_024 * 1_024
    static let copyBufferBytes = 64 * 1_024
    static let maximumDisplayNameCharacters = 255
    static let ownedStagingPrefix = "lorepia-tauri-"
    static let abandonedStagingAge: TimeInterval = 24 * 60 * 60

    static func validateReference(_ reference: String) throws {
        guard
            !reference.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
            reference.utf8.count <= maximumReferenceBytes
        else {
            throw PlatformPolicyError.invalidReference
        }
    }

    static func normalizeCredential(_ value: String) throws -> String {
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard
            !normalized.isEmpty,
            normalized.utf8.count <= maximumCredentialBytes
        else {
            throw PlatformPolicyError.invalidCredential
        }
        return normalized
    }

    static func sanitizeDisplayName(_ value: String?) -> String {
        guard let value else {
            return "selected-file"
        }
        let scalars = value.unicodeScalars.map { scalar -> Character in
            CharacterSet.controlCharacters.contains(scalar)
                ? "\u{FFFD}"
                : Character(String(scalar))
        }
        let sanitized = String(scalars.prefix(maximumDisplayNameCharacters))
        return sanitized.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            ? "selected-file"
            : sanitized
    }

    static func stagingSuffix(for displayName: String) -> String {
        switch (displayName as NSString).pathExtension.lowercased() {
        case "charx":
            ".charx"
        case "json":
            ".json"
        case "zip":
            ".zip"
        default:
            ".pending"
        }
    }

    static func shouldRemoveAbandonedStagingFile(
        name: String,
        isRegularFile: Bool,
        modifiedAt: Date?,
        now: Date
    ) -> Bool {
        guard
            name.hasPrefix(ownedStagingPrefix),
            isRegularFile,
            let modifiedAt,
            now >= modifiedAt
        else {
            return false
        }
        return now.timeIntervalSince(modifiedAt) >= abandonedStagingAge
    }
}
